sql/catalyst/src/main/java/org/apache/spark/sql/connector/write/DeltaWriter.java

```diff
@@ -0,0 +1,38 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.connector.write;
+
+import org.apache.spark.annotation.Experimental;
+
+/**
+ * A data writer returned by {@link DeltaWriterFactory#createWriter(int, long)} and is
+ * responsible for writing a delta of rows.
+ *
+ * @since 3.2.0
+ */
+@Experimental
+public interface DeltaWriter<T> extends DataWriter<T> {
+  void delete(T id);
+  void update(T id, T row);
```

Comment:
qq: is T here a "full" row? e.g. if I am going to update only some fields is that something that we could support through this API? I admit it's probably kind of an edge case but think of like a mogodb record with a large binary payload and were just changing the name?
Totally ok if not in scope for this just wanting to understand better.

Reply:
It is currently a full row but the same API can be used for passing just updated columns. We could expose an optional mix-in interface for `RowLevelOperation`.